pkg/fileutils/fileutils.go

```diff
@@ -1,26 +1,56 @@
 package fileutils
 
 import (
-	"github.com/Sirupsen/logrus"
 	"path/filepath"
+
+	"github.com/Sirupsen/logrus"
 )
 
 // Matches returns true if relFilePath matches any of the patterns
+// and isn't excluded by any of the subsequent patterns.
 func Matches(relFilePath string, patterns []string) (bool, error) {
-	for _, exclude := range patterns {
-		matched, err := filepath.Match(exclude, relFilePath)
+
+	matched := false
+
+	for _, pattern := range patterns {
+
+		var negative bool
+		if pattern == "" {
+			continue
+		}
+
+		if pattern[0] == '!' {
+			if len(pattern) == 1 {
+				continue
+			}
+			negative = true
+			pattern = pattern[1:]
+		}
```

Comment:
Yes I think ! all by itself is kind of weird because its like negating the entire .dockerignore file - so it feels like they're asking for inconsistent things.  I would either ignore it, or throw an error - I'd lean towards erroring since silently ignoring a bad situation will probably lead to more confusion.